bridges/CssSelector2Bridge.php

```diff
@@ -0,0 +1,421 @@
+     * @return string Webpage title
+     */
+    protected function getTitle($page, $title_cleanup)
+    {
+        if (is_string($page)) {
+            $page = getSimpleHTMLDOMCached($page);
+        }
+        $title = html_entity_decode($page->find('title', 0)->plaintext);
+        if (!empty($title)) {
+            $title = trim(str_replace($title_cleanup, '', $title));
+        }
+
+        return $title;
+    }
+
+    /**
+     * Remove all elements from HTML content matching cleanup selector
+     * @param string|object $content HTML content as HTML object or string
+     * @return string|object Cleaned content (same type as input)
+     */
+    protected function cleanArticleContent($content, $cleanup_selector, $remove_styling)
+    {
+        $string_convert = false;
+        if (is_string($content)) {
+            $string_convert = true;
+            $content = str_get_html($content);
+        }
+
+        if (!empty($cleanup_selector)) {
+            foreach ($content->find($cleanup_selector) as $item_to_clean) {
+                $item_to_clean->outertext = '';
+            }
+        }
+
+        if ($remove_styling) {
+            // Get rid of classes
+            $content = preg_replace('/(<[^>]+) class=".*?"/i', '$1', $content);
+            // Get rid of inline styling
+            $content = preg_replace('/(<[^>]+) style=".*?"/i', '$1', $content);
+        }
+
+        if ($string_convert) {
+            $content = $content->outertext;
+        }
+        return $content;
+    }
+
+
+    /**
+     * Retrieve first N link+element from webpage URL or DOM satisfying the specified criteria
+     * @param string|object $page URL or DOM to retrieve feed items from
+     * @param string $entry_selector DOM selector for matching HTML elements that contain article
+     *  entries
+     * @param string $url_selector DOM selector for matching links
+     * @param string $url_pattern Optional filter to keep only links matching the pattern
+     * @param int $limit Optional maximum amount of URLs to return
+     * @return array of items { <uri> => <html-element> }
+     */
+    protected function htmlFindEntryElements($page, $entry_selector, $url_selector, $url_pattern = '', $limit = 0)
+    {
+        if (is_string($page)) {
+            $page = getSimpleHTMLDOM($page);
+        }
+
+        $entryElements = $page->find($entry_selector);
+        if (empty($entryElements)) {
+            returnClientError('No entry elements for entry selector');
+        }
+
+        // Extract URIs with the associated entry element
+        $links_with_elements = [];
+        foreach ($entryElements as $entry) {
+            $url_element = $entry->find($url_selector, 0);
+            if (is_null($url_element)) {
+                // No `a` element found in this entry
+                if ($entry->tag == 'a') {
+                    $url_element = $entry;
+                } else {
+                    continue;
+                }
+            }
```

Comment:
Looks like some leftover behavior from the original `CssSelectorBridge`. Originally, when the entry selector is not a link, then the first link inside is selected for the entry URL. Here, it seems like when the user-provided selector does not match an URL inside the entry element, then if the entry element itself is an URL, it's selected instead. Not sure if that's intentional so pointing out just in case.

Reply:
It was intentional. If someone selects the parent element, but it is already an `a`, that link can be used.